Framework/API/src/DetectorInfo.cpp

```diff
@@ -0,0 +1,233 @@
+void DetectorInfo::setPosition(const Geometry::IComponent &comp,
+                               const Kernel::V3D &pos) {
+  using namespace Geometry::ComponentHelper;
+  TransformType positionType = Absolute;
+  moveComponent(comp, *m_pmap, pos, positionType);
+
+  if (!dynamic_cast<const Geometry::Detector *>(&comp)) {
+    // If comp is a detector cached positions stay valid. In all other cases
+    // (higher level in instrument tree, or other leaf component such as sample
+    // or source) we flush all cached positions.
+    if (m_source)
+      m_sourcePos = m_source->getPos();
+    if (m_sample)
+      m_samplePos = m_sample->getPos();
+    // Detector positions are currently not cached, the cached pointers to
+    // detectors stay valid. Once we store positions in DetectorInfo we need to
+    // update detector positions here.
+  }
+}
+
+/** Set the absolute rotation of the component `comp`.
+ *
+ * This may or may not be a detector. Even if it is not a detector it will
+ * typically still influence detector positions rotations. */
+void DetectorInfo::setRotation(const Geometry::IComponent &comp,
+                               const Kernel::Quat &rot) {
+  using namespace Geometry::ComponentHelper;
+  TransformType rotationType = Absolute;
+  rotateComponent(comp, *m_pmap, rot, rotationType);
+
+  if (!dynamic_cast<const Geometry::Detector *>(&comp)) {
+    // If comp is a detector cached positions and rotations stay valid. In all
+    // other cases/ (higher level in instrument tree, or other leaf component
+    // such as sample or source) we flush all cached positions and rotations.
+    if (m_source)
+      m_sourcePos = m_source->getPos();
+    if (m_sample)
+      m_samplePos = m_sample->getPos();
+    // Detector positions and rotations are currently not cached, the cached
+    // pointers to detectors stay valid. Once we store positions and rotations
+    // in DetectorInfo we need to update detector positions and rotations here.
+  }
+}
+
+/// Returns the source position.
+Kernel::V3D DetectorInfo::sourcePosition() const {
+  std::call_once(m_sourceCached, &DetectorInfo::cacheSource, this);
+  return m_sourcePos;
+}
+
+/// Returns the sample position.
+Kernel::V3D DetectorInfo::samplePosition() const {
+  std::call_once(m_sampleCached, &DetectorInfo::cacheSample, this);
+  return m_samplePos;
+}
+
+/// Returns L1 (distance from source to sample).
+double DetectorInfo::l1() const {
+  std::call_once(m_L1Cached, &DetectorInfo::cacheL1, this);
+  return m_L1;
+}
+
+/// Returns a sorted vector of all detector IDs.
+const std::vector<detid_t> DetectorInfo::detectorIDs() const {
+  return m_detectorIDs;
+}
+
+const Geometry::IDetector &DetectorInfo::getDetector(const size_t index) const {
+  size_t thread = static_cast<size_t>(PARALLEL_THREAD_NUMBER);
+  if (m_lastIndex[thread] != index) {
+    m_lastIndex[thread] = index;
+    m_detectors[thread] = m_instrument.getDetector(m_detectorIDs[index]);
```

Comment:
I would rename `m_detectors`  to `m_lastDetectors` - this makes it clearer this is for caching.